keras_nlp/utils/text_generation.py

```diff
@@ -288,63 +291,100 @@ def token_probability_fn(inputs):
     ```
 
     """
-    if not tf.executing_eagerly():
-        raise RuntimeError(
-            "`keras_nlp.utils.beam_search` currently requires an eager "
-            "execution context. Please call `beam_search` outside "
-            "tf.function or run `tf.config.run_functions_eagerly(True)` to run "
-            "tf.function in eager mode."
-        )
     if num_beams <= 0:
         raise ValueError(
             f"`num_beams` should be strictly positive. Received: `num_beams={num_beams}`."
         )
+    if num_beams == 1:
+        return greedy_search(
+            token_probability_fn=token_probability_fn,
+            prompt=prompt,
+            max_length=max_length,
+            end_token_id=end_token_id,
+            pad_token_id=pad_token_id,
+        )
 
     prompt = _validate_prompt(prompt)
 
-    if prompt.shape[-1] == 0:
-        raise ValueError(
-            "Length of `prompt` is 0, please provide a non-empty `prompt`."
-        )
-
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
-    _validate_token_probability_fn(token_probability_fn, prompt)
 
-    batch_size, length = prompt.shape
+    batch_size, length = _get_prompt_shape(prompt)
+    prompt, mask = _pad_prompt(prompt, max_length)
+
+    vocab_size = _validate_token_probability_fn(token_probability_fn, prompt)
+
     if length >= max_length:
         return tf.squeeze(prompt) if input_is_1d else prompt
 
-    # Initialize beam.
-    beams = tf.expand_dims(prompt, 1)
-    beams_prob = tf.zeros([batch_size, 1])
-    i = length
-    while i < max_length:
-        beam_size = beams.shape[1]
-        beam_preds = []
-        for j in range(beam_size):
-            preds = token_probability_fn(beams[:, j, :])
-            if from_logits:
-                preds = keras.activations.softmax(preds, axis=-1)
-            beam_preds.append(preds)
-        stacked_preds = tf.stack(beam_preds, axis=1)
-        vocab_size = stacked_preds.shape[2]
-        logits = tf.reshape(stacked_preds, [batch_size, beam_size * vocab_size])
+    # Initialize beam with shape `(batch_size, num_beams, length)`.
+    beams = tf.repeat(tf.expand_dims(prompt, axis=1), num_beams, axis=1)
+
+    # Initialize `beams_prob`` with shape `(batch_size, num_beams)`
+    beams_prob = tf.zeros([batch_size, 1], dtype=tf.float32)
+    beams_prob = tf.concat(
+        [beams_prob, tf.fill((batch_size, num_beams - 1), -3.40282e38)],
```

Comment:
Are you looking for `tf.float32.min` here?

I'm noticing you drop the softmax here, which I think is safe right `top_k(x) == top_k(softmax(x)`? Maybe we should just remove the argument for `from_logits`, same as greedy search.